[WIP] APIv4 server (#5691)
Scille/parsec-cloud: docs/rfcs/1005-realm-key-rotation.md

### 4.2 - Existing data: `UserManifest`

- Remove `last_processed_message` field.
- Deprecate `workspaces` field. The field is only kept for backward compatibility

>   as it `name` and `key` fields are needed to generate the initial realm name

Reviewer: ```suggestion
  as its `name` and `key` fields are needed to generate the initial realm name
```

  certificate and realm key rotation certificate.

> `workspace` field uses a updated `WorkspaceEntry` type:

Reviewer: ```suggestion
`workspace` field uses an updated `WorkspaceEntry` type:
```

```json5
{
  "label": "WorkspaceEntry",
  "fields": [
    {
      "name": "id",
      "type": "VlobID"
    },
    {
      "name": "name",
      "type": "EntryName"
    },
    {
        "name": "key",
        "type": "SecretKey"
    }

    // Ignored fields (useless for our backward compatibility needs)
    // - `encryption_revision`
    // - `encrypted_on`
    // - `role_cache_timestamp`
    // - `role`
  ]
}
```

> **Notes:**
>
> - In practice, this means the user manifest should never get updated as it only
>   contains data used for backward compatibility.
>   However this might change in the future (typically for storing user-specific
>   configuration should be shared between his devices).
> - The initial key rotation on a workspace created with Parsec < v3 requires to

> >   retreive the key currently in use. While the key is most likely in the last

Reviewer: ```suggestion
>   retrieve the key currently in use. While the key is most likely in the last
```

}
```

With `WorkspaceNameOrigin`:

```json5
{
    "name": "WorkspaceNameOrigin",
    "discriminant_field": "type",
    "variants": [
        // Name comes from a certificate, this is the eventualy consistent outcome.
        {
            "name": "Certificate",
            "discriminant_value": "CERTIFICATE",
            "fields": [
                {
                    "name": "timestamp",
                    "type": "DateTime"
                }
            ]
        },
        // Workspace has no realm name certificate that we are aware yet.
        // There is two reasons for that:
        // - The most likely: the workspace has just been created locally.
        // - The workspace has been shared with us before it has a realm name certificate.
        //   This theorically possible, though unlikely: it should only be encountered
        //   when dealing with a workspace shared with Parsec < v3 (as, at that time,
        //   realm name certificate didn't exist, and in Parsec >= v3 we ensure the
        //   workspace is bootstrapped before sharing it).
        //
        // In any case, this is a temporary state and the eventual outcomes are:

>         // - A realm name certificate is fetched and it name overwrites the placeholder.

Reviewer: ```suggestion
        // - A realm name certificate is fetched and its name overwrites the placeholder.
```

{
    "name": "CertificateBasedInfoOrigin",
    "discriminant_field": "type",
    "variants": [
        // Info comes from a certificate, this is the eventualy consistent outcome.
        {
            "name": "Certificate",
            "discriminant_value": "CERTIFICATE",
            "fields": [
                {
                    "name": "timestamp",
                    "type": "DateTime"
                }
            ]
        },
        // Workspace has no certificate for this info that we are aware yet.
        //
        // The most likely reason for this is that the workspace has just been created locally.
        //
        // If the info is the realm's name, another reason is that the workspace has been
        // shared with us before it has a realm name certificate.
        // This is theorically possible, though unlikely: it should only be encountered
        // when dealing with a workspace shared with Parsec < v3 (as, at that time, realm
        // name certificate didn't exist, and in Parsec >= v3 we ensure the workspace is
        // bootstrapped before sharing it).
        //
        // In any case, this is a temporary state and the eventual outcomes are:

>         // - A realm certificate is fetched and it info the placeholder.

Author: good catch, a word was missing:
> A realm certificate is fetched and it info **overwrite** the placeholder

        // bootstrapped before sharing it).
        //
        // In any case, this is a temporary state and the eventual outcomes are:

>         // - A realm certificate is fetched and it info overwrite the placeholder.

Reviewer: ```suggestion
        // - A realm certificate is fetched and its info overwrites the placeholder.
```
